tests/10apidoc/01register.pl

```diff
@@ -126,7 +128,75 @@ sub matrix_register_user
    });
 }
 
-push @EXPORT, qw( local_user_fixture local_user_fixtures );
+push @EXPORT, qw( matrix_register_user_via_secret );
+
+sub matrix_register_user_via_secret
+{
+   my ( $http, $uid, %opts ) = @_;
+
+   my $password = $opts{password} // "an0th3r s3kr1t";
+   my $admin = $opts{admin} // 0;
+
+   defined $uid or
+      croak "Require UID for matrix_register_user";
+
+   my $hmac = Digest::HMAC_SHA1->new("reg_secret");
+   $hmac->add($uid);
+   $hmac->add($password);
+   $hmac->add($admin ? "admin" : "notadmin");
+
+   my $mac = $hmac->hexdigest;
+
+   $http->do_request_json(
+      method => "POST",
+      uri    => "/api/v1/register",
+
+      content => {
+        type     => "org.matrix.login.shared_secret",
+        user     => $uid,
+        password => $password,
+        admin    => $admin ? JSON::true : JSON::false,
+        mac      => $mac,
+      },
+   )->then( sub {
+      my ( $body ) = @_;
+
+      assert_json_keys( $body, qw( user_id access_token ));
+
+      my $access_token = $body->{access_token};
+
+      my $user = User( $http, $body->{user_id}, $password, $access_token, undef, undef, undef, [], undef );
+
+      my $f = Future->done;
+
+      return $f->then_done( $user )
```

Comment:
Not sure what these two lines are doing. You're creating an already-done future, then making a new one from it with `->then_done`. Why not just have

   return Future->done( $user )
      ->on_done ...

in the first place?